subsys/bluetooth/host/smp.c

```diff
@@ -2932,7 +2932,9 @@ static uint8_t smp_pairing_req(struct bt_smp *smp, struct net_buf *buf)
 	BT_DBG("");
 
 	if ((req->max_key_size > BT_SMP_MAX_ENC_KEY_SIZE) ||
-	    (req->max_key_size < BT_SMP_MIN_ENC_KEY_SIZE)) {
+	    (req->max_key_size < BT_SMP_MIN_ENC_KEY_SIZE) ||
+	    (req->max_key_size < BT_SMP_MAX_ENC_KEY_SIZE &&
+	     IS_ENABLED(CONFIG_BT_SMP_SC_ONLY))) {
```

Comment:
Don't you need to do this in `smp_pairing_rsp()` as well? Perhaps the simplest thing would be to redefine `BT_SMP_MIN_ENC_KEY_SIZE` like this in `smp.h`:
```
#define BT_SMP_MAX_ENC_KEY_SIZE			16
#if IS_ENABLED(CONFIG_BT_SMP_SC_ONLY)
#define BT_SMP_MIN_ENC_KEY_SIZE			BT_SMP_MAX_ENC_KEY_SIZE
#else
#define BT_SMP_MIN_ENC_KEY_SIZE			7
#endif
```

Reply:
I would think so, which is why I think this handling should be moved to the existing helper function `remote_sec_level_reachable`.

```
static uint8_t remote_sec_level_reachable(struct bt_smp *smp)
{
	struct bt_conn *conn = smp->chan.chan.conn;
	bt_security_t sec;
	
	if IS_ENABLED(CONFIG_BT_SMP_SC_ONLY) {
		sec = BT_SECURITY_L4;
	else {
		sec = conn->required_sec_level;
	}

	switch (sec) {
```

This seems like the best way to fix this to me.

Reply:
I think suggestion of @joerchan is more readable, I updated a PR